### PR TITLE
feat(hooks.ts): add release hooks to create purge hooks for PubAttrib…

### PR DESCRIPTION
**Pull request by tefkah**
## Issue(s) Resolved
https://github.com/pubpub/pubpub/issues/2958

## Test Plan
Same as https://github.com/pubpub/pubpub/pull/2960, except do this also

### Added Pubs
We want to make sure a user's profile page updates if they are added to a Pub.

1. Login
2. Go to `www.duqduq.org/user/<your-user-slug>`
3. Confirm it is cached by checking for X-Cache: HIT on reload
4. Open new tab, go to a community you have no Pubs in.
5. Add yourself as an attribution to that Pub.
6. Refresh `wwww.duqduq.org/user/<your-user-slug>`, confirm X-Cache: HIT (nothing should have changed)
7. Release the Pub
8 Refresh `wwww.duqduq.org/user/<your-user-slug>`, confirm X-Cache: MISS (you should see the pub now)

## Notes/Context/Gotchas

Main changes are to make sure the comments mentioned here work: https://github.com/pubpub/pubpub/issues/2958#issuecomment-1865135864

The PR also introduces a somewhat more robust way to do purges outside of the standard middleware purge.

I define a `createPurgeHooks` function that works similarly to the `createActivityHooks` function, which works like shown here

https://github.com/pubpub/pubpub/blob/7d4774db96ef9e73bbb8ed023a48d7eb596a768c/server/pubAttribution/hooks.ts#L23-L37

Basically, you define a function that runs after a model is created/updated/destroyed, which returns either a string, and array of strings, or something nullish.
If given an array of strings or a string, it will schedule purges for those Surrogate Keys.

I moved the previously kind of haphazardly placed purges for `PUT /api/users` into such a hook, see comment below.

### CODE REVIEW

File: utils/caching/purgeMiddleware.ts [hunk 1]
```diff
@@ -126,29 +110,11 @@ export const purgeMiddleware = (
...
-					];
-
-					await Promise.all(allPurges.map(async (tag) => schedulePurge(tag)));
-				} else {
-					/**
-					 * this await is here so you can wait on the purge requests
-					 * using `finishDeferredTasks` in tests or other short lived processes
-					 */
-					await schedulePurge(surrogateTag);
-				}
```
Comment (tefkah) [inline utils/caching/purgeMiddleware.ts:151]: this was moved here: https://github.com/pubpub/pubpub/pull/2963/files#diff-091fe153990dde352868e0d97e0be2e9ebff522544644952f1348961f6cceaab

File: server/user/queries.ts [hunk 2]
```diff
@@ -103,6 +103,7 @@ export const updateUser = (
 
 	return User.update(filteredValues, {
 		where: { id: inputValues.userId },
+		individualHooks: true,
 	}).then(() => {
 		if (req.user.fullName !== filteredValues.fullName) {
 			updateUserData(req.user.id);
```
Comment (tefkah) [inline server/user/queries.ts:109]: this is necessary in order for the `User.afterUpdate` hook to run.

by default, those hooks only run for `user.update`, not `User.update({ where: { ...}})`

**Review by gabestein — APPROVED** [review]
Tested on the instance to make sure user profile stuff etc. all works and it checks out. Promoting now to check out purging on dev.
